### PR TITLE
fix: restrict allow categorical types

### PR DESCRIPTION
**Pull request by Bento007**
## Reason for Change

- during schema migration 4.0 additional categorical type combinations were found to be invalid. This PR add additional test coverage and check for those error types.
- related to https://github.com/chanzuckerberg/single-cell-curation/issues/405
## Changes

- add check for mixedd categorical types
- add check for illegal categorical types
- Add test to verify categorical types allowed.

## Testing

- unit tests.

## Notes for Reviewer

### CODE REVIEW

**Comment by Bento007** [comment]
relate to https://github.com/chanzuckerberg/single-cell-curation/issues/405

File: cellxgene_schema_cli/tests/test_validate.py [hunk 1]
```diff
@@ -355,20 +356,61 @@ def test_determine_seurat_convertibility(self):
...
+    def test_succeed_mixed_categorical_int(self, tmp_path):
+        self._succeed_categorical(tmp_path, [numpy.int32(123), int(456)])
+
+    def test_fail_categorical_bool(self, tmp_path):
+        validator = self._fail_categorical(tmp_path, [True, False])
+        assert "Categories must only contain" in validator.errors[0]
+
+    @pytest.mark.parametrize("typ", [numpy.float64, numpy.float32, float])
+    def test_fail_categorical_float64(self, tmp_path, typ):
+        validator = self._fail_categorical(tmp_path, [typ(123.123), type(456.456)])
```
Comment (nayib-jose-gloria) [inline cellxgene_schema_cli/tests/test_validate.py:375]: I can't recreate the write labels error with float64. Can you share your snippet you use to test?

I used (on an existing h5ad):
```
categories = [np.float64(4), np.float64(5)]
t = pd.CategoricalDtype(categories=categories)
adata.obs["float_cat"] = np.random.choice(categories, adata.n_obs)
adata.obs["float_cat"] = adata.obs["float_cat"].astype("category")
```

And was able to write successfully

Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/tests/test_validate.py:375]: I was also able to download the h5ads in prod that reported using floats and I was able to read and write on them

Comment (Bento007) [reply cellxgene_schema_cli/tests/test_validate.py:375]: I tried your way and it does work. When I use the methods:

```
        categories = [np.float32(4), np.float32(5)]
        t = pd.CategoricalDtype(categories=categories)
        adata.obs["not_string"] = pd.Series(data=categories, index=["X", "Y"], dtype=t)
```

I end up with two class types in the category. A `float` class type and `type` class type. I have no idea where the `type` class type came from. None of the other tests cases do this and the other tests are using the same method.

OMG I found it, it's a typo on my end. >> `type(456.456)`->`typ(456.456)``

File: cellxgene_schema_cli/cellxgene_schema/validate.py [hunk 2]
```diff
@@ -707,12 +707,22 @@ def _validate_dataframe(self, df_name: str):
...
                     self.errors.append(
-                        f"Column '{column_name}' in dataframe '{df_name}' must only contain string categories. Found {category_types}."
+                        f"Column '{column_name}' in dataframe '{df_name}' containes {illegal_categorical_types=}."
+                    )
+                # Check for categorical column has mixed types, which is not supported by anndata 0.8.0
+                # TODO: check if this can be removed after upgading to anndata 0.10.0
+                categorical_types = {type(x) for x in column.dtype.categories.values}
+                if len(categorical_types) > 1:
+                    self.errors.append(
+                        f"Column '{column_name}' in dataframe '{df_name}' containes {len(categorical_types)} categorical types. "
```
Comment (nayib-jose-gloria) [inline cellxgene_schema_cli/cellxgene_schema/validate.py:724]:
```suggestion
                        f"Column '{column_name}' in dataframe '{df_name}' contains {len(categorical_types)} categorical types. "
```

File: cellxgene_schema_cli/tests/test_validate.py [hunk 3]
```diff
@@ -355,35 +360,74 @@ def test_determine_seurat_convertibility(self):
...
         validator._validate_dataframe("obs")
-        assert "must only contain string categories." in validator.errors[0]
-        with pytest.raises(TypeError):
-            # If this tests starts to fail here it means the anndata version has be upgraded and this check is no
-            # longer needed
-            adata.write_h5ad(f"{tmp_path}/test.h5ad")
 
-    def test_fail_mixed_column_types(self, tmp_path):
+        # Assert
+        assert "in dataframe 'obs' containes 2 categorical types. Only one type is allowed." in validator.errors[0]
```
Comment (nayib-jose-gloria) [inline cellxgene_schema_cli/tests/test_validate.py:387]:
```suggestion
        assert "in dataframe 'obs' contains 2 categorical types. Only one type is allowed." in validator.errors[0]
```

File: cellxgene_schema_cli/cellxgene_schema/validate.py [hunk 4]
```diff
@@ -707,12 +707,22 @@ def _validate_dataframe(self, df_name: str):
...
+                # Check for columns that have illegal categories, which are not supported by anndata 0.8.0
                 # TODO: check if this can be removed after upgading to anndata 0.10.0
-                category_types = {type(x) for x in column.dtype.categories.values}
-                if len(category_types) > 1 or str not in category_types:
+                blocked_categorical_types = {bool}
+                illegal_categorical_types = categorical_types & blocked_categorical_types
+                if illegal_categorical_types:
                     self.errors.append(
-                        f"Column '{column_name}' in dataframe '{df_name}' must only contain string categories. Found {category_types}."
+                        f"Column '{column_name}' in dataframe '{df_name}' containes {illegal_categorical_types=}."
```
[the remark this answers is not in the document]
Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/cellxgene_schema/validate.py:717]: I did not know about that equal sign feature in f-strings, that's cool

**Review by nayib-jose-gloria — APPROVED** [review]
there's a typo to fix but the change looks good, ty!

File: cellxgene_schema_cli/tests/test_validate.py [hunk 5]
```diff
@@ -355,35 +360,74 @@ def test_determine_seurat_convertibility(self):
...
+        # Arrange
+        categories = [True, False]
+        self._add_catagorical_obs(valid_adata, categories)
+        validator = self._create_validator(valid_adata)
+
+        # Act
+        validator._validate_dataframe("obs")
+
+        # Assert
+        assert "in dataframe 'obs' containes illegal_categorical_types={<class 'bool'>}." in validator.errors[0]
```
Comment (Bento007) [inline cellxgene_schema_cli/tests/test_validate.py:400]:
```suggestion
        assert "in dataframe 'obs' contains illegal_categorical_types={<class 'bool'>}." in validator.errors[0]
```

**Pull request merged**